rfcs/014-pagination.md

```diff
@@ -0,0 +1,193 @@
+- _all_dbs - the endpoint returns a list instead of object as the rest of the endpoints
+- _changes - the endpoint contains too many different modes and it would require more careful consideration
+
+The endpoints in scope are:
+- {db}/_all_docs
+- {db}/_all_docs/queries
+- _dbs_info
+- {db}/_design/{ddoc}/_view/{view}
+- {db}/_design/{ddoc}/_view/{view}/queries
+
+In a nutshell the idea is:
+- add `page_size` query field to control number of rows on each page and to flag
+  that client expects paginated response
+- add `first`, `previous`, `next` fields which contain bookmark part of URI
+- add `bookmark` query field to retrieve bookmarked page
+
+
+## Requirements Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+- `bookmark` - is opaque token which would contain information needed to retrieve
+  bookmarked page. The format of the token value MUST NOT be relied upon in the client.
+
+---
+
+# Detailed Description
+
+## Implementation proposal
+
+1) Add new optional query field called `bookmark` to following endpoints
+  - {db}/_all_docs
+  - {db}/_all_docs/queries
+  - _dbs_info
+  - {db}/_design/{ddoc}/_view/{view}
+  - {db}/_design/{ddoc}/_view/{view}/queries
+
+2) Add new optional query field called `page_size` if it is set we would use
+  paginated endpoint otherwise use old code path
+3) Implement per-endpoint configurable max limits to restrict page size of
+  paginated responses
+
+ ```
+ [request_limits]
+ _all_docs = 5000
+ _all_docs/queries = 5000
+ _all_dbs = 5000
+ _dbs_info = 5000
+ _view = 2500
+ _view/queries = 2500
+ _find = 2500
+ ```
+4) Add following additional fields into response:
+  ```
+    "first": "12345678945621321689",
+    "previous": "983uiwfjkdsdf",
+    "next": "12343tyekf3"
+  ```
+
+## Limitations
+
+- The `first`/`next`/`last` keys in the response are represented as path which
+  includes the bookmark query key. This means the bookmark token size contributes
+  to total URI length and is subject to a max URL lenght (8192 these days).
+  CouchDB historically didn't have limit on docid. FDB imposes a limit on a key path.
+  Therefore we would probably have a strict limit in CouchDB 4.0. We didn't agree
+  on the maximum length of docid yet. However it seems to be trending towards 5kB
+  (see [RFC 008's comments](https://github.com/apache/couchdb-documentation/pull/410#discussion_r340536640)).
+  With 5kB sized keys we would exceed 8192 characters when both `start_key` and
+  `end_key` would be specified. This means we would have to support POST requests.
+
+## Semantics of the implementation
+
+- The bookmark would include information needed to ensure proper pagination
+  without the need to repeat initial parameters of the request.
```

Comment:
Which fields would it include and how would they change from request to request?

Reply:
The PoC would include all non default values of mrargs record. However it is an implementation detail and subject to change. The plan is to update only three fields:
- `start_key` (or `start_key_docid`)
- `page_size` we might need to half it on timeouts (we might have to store initial value as well)
- `keys` we can remove already return keys (optimization)

Reply:
What about the `include_docs=true|false` field, that is part of mrargs, would users need to provide it along with each bookmark eventually?

Reply:
> What about the include_docs=true|false field, that is part of mrargs, would users need to provide it along with each bookmark eventually?

I think it is sufficient to provide it once.